### ISSUE TITLE
Compiling 2.2 fails under macOS with error: no member named 'st_atim' in 'struct stat'

### ISSUE THREAD

**Issue by bFanek** [rpodgorny/unionfs-fuse#101]
I am running latest macOS (Big Sur) on Intel Mac.

I can compile version 2.1 without any issues.

However, version 2.2 fails when I compile:

> [  5%] Building C object src/CMakeFiles/unionfsctl.dir/unionfsctl.o
> [ 11%] Linking C executable unionfsctl
> [ 11%] Built target unionfsctl
> [ 16%] Building C object src/CMakeFiles/unionfs.dir/unionfs.o
> [ 22%] Building C object src/CMakeFiles/unionfs.dir/opts.o
> [ 27%] Building C object src/CMakeFiles/unionfs.dir/debug.o
> [ 33%] Building C object src/CMakeFiles/unionfs.dir/findbranch.o
> [ 38%] Building C object src/CMakeFiles/unionfs.dir/readdir.o
> [ 44%] Building C object src/CMakeFiles/unionfs.dir/general.o
> [ 50%] Building C object src/CMakeFiles/unionfs.dir/unlink.o
> [ 55%] Building C object src/CMakeFiles/unionfs.dir/cow.o
> In file included from /Users/b/Downloads/unionfs-fuse-2.2/src/cow.c:24:
> /Users/b/Downloads/unionfs-fuse-2.2/src/cow_utils.h:11:9: warning: 'MAXBSIZE' macro redefined [-Wmacro-redefined]
> #define MAXBSIZE 4096
>         ^
> /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.3.sdk/usr/include/sys/param.h:181:9: note: previous definition is here
> #define MAXBSIZE        (256 * 4096)
>         ^
> 1 warning generated.
> [ 61%] Building C object src/CMakeFiles/unionfs.dir/cow_utils.o
> /Users/b/Downloads/unionfs-fuse-2.2/src/cow_utils.c:68:14: error: no member named 'st_atim' in 'struct stat'
>         ut[0] = fs->st_atim;
>                 ~~  ^
> /Users/b/Downloads/unionfs-fuse-2.2/src/cow_utils.c:69:14: error: no member named 'st_mtim' in 'struct stat'
>         ut[1] = fs->st_mtim;
>                 ~~  ^
> 2 errors generated.
> make[2]: *** [src/CMakeFiles/unionfs.dir/cow_utils.o] Error 1
> make[1]: *** [src/CMakeFiles/unionfs.dir/all] Error 2
> make: *** [all] Error 2

Any clue?

Cheers!

**Comment by TheKnarf** [comment]
I'm having the same build issue.

Here's some info about my system:

```
$ npx envinfo --system --utilities

  System:
    OS: macOS 12.0.1
    CPU: (8) arm64 Apple M1
    Memory: 103.50 MB / 16.00 GB
    Shell: 5.8 - /bin/zsh
  Utilities:
    CMake: 3.22.0 - /opt/homebrew/bin//cmake
    Make: 3.81 - /usr/bin/make
    GCC: 4.2.1 - /usr/bin/gcc
    Git: 2.34.0 - /opt/homebrew/bin//git
    Clang: 13.0.0 - /usr/bin/clang
    Subversion: 1.14.1 - /opt/homebrew/bin//svn
```

**Comment by TheKnarf** [comment]
The CMakeFile refers to `osxfuse` on the current line:

https://github.com/rpodgorny/unionfs-fuse/blob/master/CMakeLists.txt#L19

**Comment by rpodgorny** [comment]
can you, please, try with the current master? i've reverted the old stat() behaviour for macos.

unfortunately, i have no mac system to test myself...

**Comment by rpodgorny** [comment]
> The CMakeFile refers to `osxfuse` on the current line:
> 
> https://github.com/rpodgorny/unionfs-fuse/blob/master/CMakeLists.txt#L19
> 
> However I belive `osxfuse` is deprecated and replaced with `macfuse`.

feel free to change it and try and report back to me - i have no way to test it so i'm relying on contributed patches... :-(

**Comment by TheKnarf** [comment]
Building with CMake does not seem to work on MacOS from master (`b6b02ae`)

```
$  mkdir build && cd build
$ cmake ..
-- The C compiler identification is AppleClang 13.0.0.13000029
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Library/Developer/CommandLineTools/usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
CMake Warning (dev) at CMakeLists.txt:3 (cmake_minimum_required):
  Compatibility with CMake < 2.4 is not supported by CMake >= 3.0.
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Deprecation Warning at CMakeLists.txt:3 (cmake_minimum_required):
  Compatibility with CMake < 2.8.12 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


-- Configuring done
-- Generating done
-- Build files have been written to: /Users/knarf/projects/unionfs-fuse/build
$ make
[  5%] Building C object src/CMakeFiles/unionfs.dir/unionfs.o
/Users/knarf/projects/unionfs-fuse/src/unionfs.c:13:10: fatal error: 'fuse.h' file not found
#include <fuse.h>
         ^~~~~~~~
1 error generated.
make[2]: *** [src/CMakeFiles/unionfs.dir/unionfs.o] Error 1
make[1]: *** [src/CMakeFiles/unionfs.dir/all] Error 2
make: *** [all] Error 2
```

```
npx envinfo --system --utilities

  System:
    OS: macOS 12.0.1
    CPU: (8) arm64 Apple M1
    Memory: 80.11 MB / 16.00 GB
    Shell: 5.8 - /bin/zsh
  Utilities:
    CMake: 3.22.0 - /opt/homebrew/bin//cmake
    Make: 3.81 - /usr/bin/make
    GCC: 4.2.1 - /usr/bin/gcc
    Git: 2.34.0 - /opt/homebrew/bin//git
    Clang: 13.0.0 - /usr/bin/clang
    Subversion: 1.14.1 - /opt/homebrew/bin//svn
```

**Comment by TheKnarf** [comment]
After updating the include path as following:

```
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 093cd36..5b4a256 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -16,7 +16,10 @@ SET(CMAKE_C_FLAGS_RELEASE "-O2")
 SET(CMAKE_C_FLAGS_DEBUG "-O0 -g -DDEBUG")

 if (UNIX AND APPLE)
+       # Legacy include path for osxfuse
   include_directories("/usr/local/include/osxfuse/fuse")
+       # New include path for MacFuse
+  include_directories("/usr/local/include/fuse")
 endif()

 add_definitions(-D_FILE_OFFSET_BITS=64 -DFUSE_USE_VERSION=26)
```

It seems to build but fail at the linker step:

```
make
[  5%] Building C object src/CMakeFiles/unionfs.dir/unionfs.o
[ 11%] Building C object src/CMakeFiles/unionfs.dir/opts.o
[ 16%] Building C object src/CMakeFiles/unionfs.dir/debug.o
[ 22%] Building C object src/CMakeFiles/unionfs.dir/findbranch.o
[ 27%] Building C object src/CMakeFiles/unionfs.dir/readdir.o
[ 33%] Building C object src/CMakeFiles/unionfs.dir/general.o
[ 38%] Building C object src/CMakeFiles/unionfs.dir/unlink.o
[ 44%] Building C object src/CMakeFiles/unionfs.dir/cow.o
In file included from /Users/knarf/projects/unionfs-fuse/src/cow.c:24:
/Users/knarf/projects/unionfs-fuse/src/cow_utils.h:11:9: warning: 'MAXBSIZE' macro redefined [-Wmacro-redefined]
#define MAXBSIZE 4096
        ^
/Library/Developer/CommandLineTools/SDKs/MacOSX12.0.sdk/usr/include/sys/param.h:181:9: note: previous definition is here
#define MAXBSIZE        (256 * 4096)
        ^
1 warning generated.
[ 50%] Building C object src/CMakeFiles/unionfs.dir/cow_utils.o
[ 55%] Building C object src/CMakeFiles/unionfs.dir/string.o
[ 61%] Building C object src/CMakeFiles/unionfs.dir/rmdir.o
[ 66%] Building C object src/CMakeFiles/unionfs.dir/usyslog.o
/Users/knarf/projects/unionfs-fuse/src/usyslog.c:160:17: warning: overlapping comparisons always evaluate to false [-Wtautological-overlap-compare]
        if (tinfo == 0 && tinfo == 1)
            ~~~~~~~~~~~^~~~~~~~~~~~~
1 warning generated.
[ 72%] Building C object src/CMakeFiles/unionfs.dir/fuse_ops.o
/Users/knarf/projects/unionfs-fuse/src/fuse_ops.c:129:38: warning: unused parameter 'path' [-Wunused-parameter]
static int unionfs_flush(const char *path, struct fuse_file_info *fi) {
                                     ^
/Users/knarf/projects/unionfs-fuse/src/fuse_ops.c:150:38: warning: unused parameter 'path' [-Wunused-parameter]
static int unionfs_fsync(const char *path, int isdatasync, struct fuse_file_info *fi) {
                                     ^
/Users/knarf/projects/unionfs-fuse/src/fuse_ops.c:398:37: warning: unused parameter 'path' [-Wunused-parameter]
static int unionfs_read(const char *path, char *buf, size_t size, off_t offset, struct fuse_file_info *fi) {
                                    ^
/Users/knarf/projects/unionfs-fuse/src/fuse_ops.c:426:40: warning: unused parameter 'path' [-Wunused-parameter]
static int unionfs_release(const char *path, struct fuse_file_info *fi) {
                                       ^
4 warnings generated.
[ 77%] Building C object src/CMakeFiles/unionfs.dir/hashtable.o
[ 83%] Building C object src/CMakeFiles/unionfs.dir/hashtable_itr.o
[ 88%] Linking C executable unionfs
ld: library not found for -lfuse
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [src/unionfs] Error 1
make[1]: *** [src/CMakeFiles/unionfs.dir/all] Error 2
make: *** [all] Error 2
```

**Comment by TheKnarf** [comment]
Seems like by adding `link_directories('/usr/local/lib')` it successfully builds on my Mac. #105.

Now I just need to test that it works.

**Comment by rpodgorny** [comment]
> Seems like by adding `link_directories('/usr/local/lib')` it successfully builds on my Mac. #105.
> 
> Now I just need to test that it works.

btw, don't forget there's an automated test suite (just run "make test" - python needed)

**Comment by rpodgorny** [comment]
i've just updated the master branch with some cmake updates - plase be sure to test with the new setup. thanks!

**Comment by rpodgorny** [comment]
this should be solved in the upcoming release. please test the current master branch if you find some time, thanks...

#110

**Comment by rpodgorny** [comment]
so, i've just released version 3.0 which should have this fixed. if not, feel free to reopen...
